database/abusescanner.go

```diff
@@ -0,0 +1,429 @@
+			allBlocked = false
+		}
+		switch a.BlockResult[i] {
+		case AbuseStatusBlocked:
+		default:
+		}
+		sb.WriteString(fmt.Sprintf("%s\n", strings.Join(parts, " | ")))
+	}
+
+	sb.WriteString("\nSummary:\n")
+	if len(pr.Skylinks) == 0 {
+		sb.WriteString("FAILURE - no skylinks found.\n")
+	} else {
+		if allBlocked {
+			sb.WriteString("SUCCESS\n")
+		} else {
+			sb.WriteString("FAILURE - not all skylinks blocked.\n")
+		}
+	}
+
+	return sb.String()
+}
+
+// NewAbuseScannerDB returns an instance of the Mongo DB.
+func NewAbuseScannerDB(connectionString, portalHostName string, logger *logrus.Logger) (*AbuseScannerDB, error) {
+	// create the client
+	opts := options.Client().ApplyURI(connectionString)
+	client, err := mongo.NewClient(opts)
+	if err != nil {
+		return nil, err
+	}
+
+	// create a context with timeout
+	ctx, cancel := context.WithTimeout(context.Background(), mongoDefaultTimeout)
+	defer cancel()
+
+	// connect to the client
+	err = client.Connect(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	// get a database handler
+	database := client.Database(dbName)
+
+	// ensure the locks collection
+	if database.Collection(collLocks) == nil {
+		err = database.CreateCollection(ctx, collLocks)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	// create the mongo database
+	db := &AbuseScannerDB{
+		MongoDB{
+			staticClient:   client,
+			staticDatabase: database,
+			staticLogger:   logger,
+			staticName:     dbName,
+		},
+		*lock.NewClient(database.Collection(collLocks)),
+		portalHostName,
+	}
+
+	// the lock client creates its own indices
+	err = db.CreateIndexes(ctx)
+	if err != nil {
+		return nil, errors.AddContext(err, "failed to create indices on locks")
+	}
+
+	// ensure the schema
+	err = db.ensureSchema(ctx, map[string][]mongo.IndexModel{
+		collEmails: {
+			{
+				Keys:    bson.D{{"email_uid", 1}},
+				Options: options.Index().SetUnique(true),
+			},
+			{
+				Keys:    bson.D{{"parsed", 1}},
+				Options: options.Index(),
+			},
+			{
+				Keys:    bson.D{{"blocked", 1}},
+				Options: options.Index(),
+			},
+			{
+				Keys:    bson.D{{"finalized", 1}},
+				Options: options.Index(),
+			},
+		},
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return db, nil
+}
+
+// FindOne returns the message with given uid
+func (db *AbuseScannerDB) FindOne(emailUid string) (*AbuseEmail, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), mongoDefaultTimeout)
+	defer cancel()
+
+	collEmails := db.staticDatabase.Collection(collEmails)
+	res := collEmails.FindOne(ctx, bson.M{"email_uid": emailUid})
+	if isDocumentNotFound(res.Err()) {
+		return nil, nil
+	}
+	if res.Err() != nil {
+		return nil, res.Err()
+	}
+
+	var email AbuseEmail
+	err := res.Decode(&email)
+	if err != nil {
+		return nil, err
+	}
+	return &email, nil
+}
+
+// FindUnblocked returns the messages that have not been blocked.
+func (db *AbuseScannerDB) FindUnblocked() ([]AbuseEmail, error) {
```

Comment:
nit: `FindUnblocked`, `FindUnfinalized` and `FindUnparsed` are basically the same method with a different filter. We can DRY this at some point.

Reply:
yeah makes sense, added `findGeneric`